### ISSUE TITLE
class

### ISSUE THREAD

**Issue by luocmin** [mseg-dataset/mseg-semantic#15]
There's a little bit of an error with the class
![image](https://user-images.githubusercontent.com/57932842/96395494-58d6d280-11f7-11eb-9d64-023833bc2deb.png)

**Comment by luocmin** [comment]
Could you please tell me how to fix this error? And where did I find the args.dataset_name.
![image](https://user-images.githubusercontent.com/57932842/96396688-54f87f80-11fa-11eb-9f80-8f5be70f07d2.png)

**Comment by luocmin** [comment]
Author, there is this problem at present, I don't know how to solve it, the situation is urgent, thank you！！

**Comment by luocmin** [comment]
This is how I changed the number of dataset into the number of GPU 1. I don't know if this means the training has begun
![image](https://user-images.githubusercontent.com/57932842/96408276-bda12580-1215-11eb-88b9-a37277794894.png)

**Comment by luocmin** [comment]
Author, could you spare me a moment? Could you help me with the question I had two days ago? Thank you

**Comment by johnwlambert** [comment]
Hi @luocmin , please pull the latest version. Let me know if this doesn't answer your questions:

(1) You're right that `tc.classes` has been changed to `tc.num_uclasses`, thanks for catching that. I've corrected the `train.py` script in my latest commit.

(2) `dataset_name` is set in Line 497: https://github.com/mseg-dataset/mseg-semantic/blob/training/mseg_semantic/tool/train.py#L497

(3) Please pull in the latest `master` of `mseg-semantic` into your branch to see the parameter `use_naive_taxonomy`: https://github.com/mseg-dataset/mseg-semantic/blob/master/mseg_semantic/utils/transform.py#L54

(4) I didn't catch how you are doing the dataset to GPU mapping, could you explain in more detail here?

If you are limited by GPU RAM, you could also

**Comment by luocmin** [comment]
@johnwlambert  Hi,
1、What I want to ask about this picture is what is the interface for successful training?
![image](https://user-images.githubusercontent.com/57932842/96820862-6be8dd00-1459-11eb-9854-96e2ec137685.png)

2、In addition, I modified some of the code in the MSEg-3m.YAMl file, 
May I ask whether this modification is correct ？
as shown below：
change:
![image](https://user-images.githubusercontent.com/57932842/96820950-9470d700-1459-11eb-837f-fdd0005f6a83.png)
to:dataset: [ade20k-150-relabeled]
GPU_map:
![image](https://user-images.githubusercontent.com/57932842/96821011-b2d6d280-1459-11eb-83d9-95368785433f.png)
to:dataset_gpu_mapping: {'ade20k-150-relabeled': [0]}

**Comment by luocmin** [comment]
3、If I modified this class, do I need to recompile mseg_semantic
![image](https://user-images.githubusercontent.com/57932842/96822436-a56f1780-145c-11eb-9a15-e61dfbf16b8e.png)

**Comment by luocmin** [comment]
The following problems occurred in training two data sets with two CARDS. What is the reason? How to solve it?
![image](https://user-images.githubusercontent.com/57932842/96831294-60081580-146f-11eb-97b9-3d8a9c8203ea.png)

**Comment by johnwlambert** [comment]
Hi @luocmin, there is no compilation involved in our repo since all files are pure Python or bash.

Our configuration is to use 7 processes. Each process processes one dataset
https://github.com/mseg-dataset/mseg-semantic/blob/f8afb3cb637bd5e921a1689681e5a7044a716b57/mseg_semantic/tool/train.py#L537

The gpu index (0,1,2...,6) is the `rank` and we call

```python
def get_rank_to_dataset_map(args) -> Dict[int,str]:
    """
        Obtain a mapping from GPU rank (index) to the name of the dataset residing on this GPU.
        Args:
        -   args
        Returns:
        -   rank_to_dataset_map
    """
    rank_to_dataset_map = {}
    for dataset, gpu_idxs in args.dataset_gpu_mapping.items():
        for gpu_idx in gpu_idxs:
            rank_to_dataset_map[gpu_idx] = dataset
    print('Rank to dataset map: ', rank_to_dataset_map)
    return rank_to_dataset_map

args.dataset_name = rank_to_dataset_map[args.rank]
...
train_data = dataset.SemData(split='train', data_root=args.data_root[args.dataset_name], data_list=args.train_list[args.dataset_name], transform=train_transform)
...
train_loader = torch.utils.data.DataLoader(train_data, batch_size=args.batch_size, shuffle=(train_sampler is None), num_workers=args.workers, pin_memory=True, sampler=train_sampler, drop_last=True)
```
See [here](https://github.com/mseg-dataset/mseg-semantic/blob/f8afb3cb637bd5e921a1689681e5a7044a716b57/mseg_semantic/tool/train.py#L497)

Changing our config by mapping each dataset to the same GPU will mean that only one dataset is trained (the last one to hit line 394 https://github.com/mseg-dataset/mseg-semantic/blob/f8afb3cb637bd5e921a1689681e5a7044a716b57/mseg_semantic/tool/train.py#L394).

You will need a different strategy about how to use fewer GPUs. I mentioned a few already (concatenating all image IDs into a single dataset, which could be sharded across 4 gpus, or instead you could accumulate gradients in place over 2 forward and backward passes, and then perform a single gradient update).

**Comment by luocmin** [comment]
Thank you, but being white I won't change the dataloader code for now

**Comment by luocmin** [comment]
I have been trying to run two data sets with two CARDS, but it still doesn't work. May I ask if I need to use the script you mentioned before to run the command? There is no script in the report.
![image](https://user-images.githubusercontent.com/57932842/96948530-76b67700-1518-11eb-90b7-51cafd144ce2.png)
This is the command I ran on the server with an error:
![image](https://user-images.githubusercontent.com/57932842/96948891-33103d00-1519-11eb-9960-028bbe61b690.png)
